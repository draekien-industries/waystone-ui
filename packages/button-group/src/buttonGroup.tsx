'use client';

import { type MouseEvent as ReactMouseEvent } from 'react';
import { Button, type ButtonProps, type ButtonSize } from '@waystone/button';
import { Flex } from 'theme-ui';

export type ButtonGroupItemProps = Omit<
  ButtonProps,
  | 'active'
  | 'size'
  | 'variant'
  | 'color'
  | 'fullWidth'
  | 'loading'
  | 'id'
  | 'key'
  | 'button'
>;

export type ButtonGroupItem<TIdentifier extends string = string> = {
  id: TIdentifier;
} & ButtonGroupItemProps;

export type ButtonGroupProps<TIdentifier extends string = string> = {
  /** The ID of the button that is currently selected */
  selected?: TIdentifier;
  /**
   * The collection of buttons to render inside the group.
   * A unique collection is extracted using the `id` of each button.
   */
  buttons: Array<ButtonGroupItem<TIdentifier>>;
  /** The callback that will be invoked when the selected button changes. */
  onChange?: (
    selected: NoInfer<TIdentifier>,
    e: ReactMouseEvent<HTMLButtonElement, MouseEvent>
  ) => void;
  /** The size of the button group */
  size?: ButtonSize;
};

/**
 * Renders a group of buttons where only one button is able to be selected at the same time
 * @param param0
 * @returns
 */
export const ButtonGroup = <TIdentifier extends string = string>({
  buttons = [],
  selected,
  onChange,
  size = 'small',
}: ButtonGroupProps<TIdentifier>) => {
  if (!buttons.length) return null;

  const uniqueButtons = [
    ...new Map(
      buttons.filter(Boolean).map((button) => [button.id, button])
    ).values(),
  ];

  return (
    <Flex
      sx={{
        gap: 'small',
        padding: 'extra-small',
        borderRadius: 'small',
        backgroundColor: 'ghost',
        width: 'fit-content',
      }}>
      {uniqueButtons.map(({ id, ...rest }) => (
        <Button
          id={id}
          key={id}
          active={id === selected}
          variant="ghost"
          size={size}
          type="button"
          onClick={(e) => {
            if (onChange) onChange(id, e);
          }}
          {...rest}
        />
      ))}
    </Flex>
  );
};
