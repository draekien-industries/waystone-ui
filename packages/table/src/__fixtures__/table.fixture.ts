import { faker } from '@faker-js/faker';
import type { ColumnDef } from '@tanstack/react-table';
import { createInteractiveColumn } from '../components/prefabs';

type Person = {
  firstName: string;
  lastName: string;
  age: number;
  visits: number;
  progress: number;
  status: 'relationship' | 'complicated' | 'single';
  subRows?: Person[];
};

const range = (len: number) => {
  const arr: number[] = [];
  for (let i = 0; i < len; i++) {
    arr.push(i);
  }
  return arr;
};

faker.seed(459201);

const newPerson = (): Person => ({
  firstName: faker.person.firstName(),
  lastName: faker.person.lastName(),
  age: faker.number.int(40),
  visits: faker.number.int(1000),
  progress: faker.number.int(100),
  status: faker.helpers.shuffle<Person['status']>([
    'relationship',
    'complicated',
    'single',
  ])[0]!,
});

function makeData(...lens: number[]) {
  const makeDataLevel = (depth = 0): Person[] => {
    const len = lens[depth]!;
    return range(len).map(
      (): Person => ({
        ...newPerson(),
        subRows: lens[depth + 1] ? makeDataLevel(depth + 1) : undefined,
      })
    );
  };

  return makeDataLevel();
}

export const columns: ColumnDef<Person>[] = [
  createInteractiveColumn<Person>(),
  {
    accessorKey: 'firstName',
    header: 'First Name',
  },
  {
    accessorKey: 'lastName',
    header: 'Last Name',
  },
  {
    accessorKey: 'age',
    header: 'Age',
  },
  {
    accessorKey: 'visits',
    header: 'Visits',
  },
  {
    accessorKey: 'status',
    header: 'Status',
  },
  {
    accessorKey: 'progress',
    header: 'Progress',
  },
];

export const data = makeData(20, 5, 3);
