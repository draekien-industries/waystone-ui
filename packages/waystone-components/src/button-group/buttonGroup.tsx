'use client';

import { Flex } from 'theme-ui';
import { Button, ButtonProps } from '../button';
import { buttonGroupCss } from './buttonGroup.styles';

/** The props provided to the onClick callback */
export type onClickProps = {
  /** The raw react mouse event. */
  e: React.MouseEvent<HTMLButtonElement, MouseEvent>;
  /** The button that was clicked. */
  clicked: string;
};

/** The props for the button group component. */
export interface ButtonGroupProps {
  /** The button that is currently active. */
  activeButton?: string;
  /** The buttons to be rendered. */
  buttons: Pick<ButtonProps, 'id' | 'children' | 'icon'>[];
  /** The callback function that will be called when the active button changes. */
  onChange?: (props: onClickProps) => void;
}

/**
 * Renders a button group component which allows one button to be selected at a time.
 * @param props - the {@link ButtonGroupProps}
 * @returns the button group component
 */
export const ButtonGroup = ({
  activeButton,
  buttons,
  onChange,
}: ButtonGroupProps) => (
  <Flex sx={buttonGroupCss}>
    {buttons.map((button) => {
      if (!button.id) {
        throw new Error('You must provide an `id` for each button');
      }

      const key = button.id;

      return (
        <Button
          {...button}
          key={key}
          variant="ghost"
          active={key === activeButton}
          onClick={(e) => {
            if (onChange) onChange({ e, clicked: key });
          }}
        />
      );
    })}
  </Flex>
);
