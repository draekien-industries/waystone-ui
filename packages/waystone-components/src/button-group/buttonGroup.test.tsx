import { screen, render, fireEvent } from '../__fixtures__';
import { ButtonGroup } from './buttonGroup';

describe('The ButtonGroup component', () => {
  it('should match snapshot', () => {
    const { container } = render(
      <ButtonGroup
        buttons={[
          { id: 'button1', children: 'button1' },
          { id: 'button2', children: 'button2' },
        ]}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it('should throw an error when no ids are provided for buttons', () => {
    expect(() =>
      render(
        <ButtonGroup
          buttons={[
            {
              children: 'button',
            },
          ]}
        />
      )
    ).toThrowError('You must provide an `id` for each button');
  });

  it('invokes `onChange` when a button is clicked', () => {
    const handleChange = jest.fn();
    render(
      <ButtonGroup
        buttons={[
          { id: 'button1', children: 'button1' },
          { id: 'button2', children: 'button2' },
        ]}
        onChange={handleChange}
      />
    );

    const button1 = screen.getByText('button1');

    fireEvent.click(button1);

    expect(handleChange).toBeCalled();
  });

  it('renders the correct active button', () => {
    render(
      <ButtonGroup
        buttons={[{ id: 'button1', children: 'button1' }]}
        activeButton="button1"
      />
    );

    const button1 = screen.getByRole('button');

    expect(button1).toHaveStyle({
      outlineStyle: 'solid',
    });
  });
});
