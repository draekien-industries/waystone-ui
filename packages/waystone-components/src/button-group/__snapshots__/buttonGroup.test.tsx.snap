// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ButtonGroup component should match snapshot 1`] = `
<div>
  <div
    class="css-ofnuv4-ButtonGroup"
  >
    <button
      class="css-4eztqb-Button"
      id="button1"
    >
      <span
        class="css-3irkks-Text"
      >
        button1
      </span>
    </button>
    <button
      class="css-4eztqb-Button"
      id="button2"
    >
      <span
        class="css-3irkks-Text"
      >
        button2
      </span>
    </button>
  </div>
</div>
`;
